tiledb/ndrectangle.py

```diff
@@ -0,0 +1,64 @@
+from typing import Tuple, Union
+
+import tiledb
+import tiledb.cc as lt
+
+from .ctx import Ctx, CtxMixin
+from .domain import Domain
+
+
+class NDRectangle(CtxMixin, lt.NDRectangle):
+    """
+    Represents a TileDB N-Dimensional Rectangle.
+    """
+
+    _dim_types = {}
+
+    def __init__(self, ctx: Ctx, domain: Domain):
+        """Class representing an N-Dimensional Rectangle of a TileDB Domain.
+
+        :param ctx: A TileDB Context
+        :param domain: A TileDB Domain
+        :raises tiledb.TileDBError:
+        """
+        super().__init__(ctx, domain)
+
+    def set_range(
+        self,
+        dim: Union[str, int],
+        start: Union[int, float, str],
+        end: Union[int, float, str],
+    ):
+        """Sets a range for the given dimension.
+
+        :param dim: Dimension name or index
+        :param start: Range start value
+        :param end: Range end value
+        :raises tiledb.TileDBError:
+        """
+        # Set types only if range is set successfully
+        try:
+            self._set_range(dim, start, end)
+            self._dim_types[dim] = type(start).__name__
+        except Exception as e:
+            raise e
```

Comment:
The try-except block here is unnecessary as we don't add additional handling or information to the exception.